### DISCUSSION TITLE
One more argument for supporting something like middleware in StateAdapt

### DISCUSSION THREAD

**Discussion by Evertt** [state-adapt/state-adapt#76] [Ideas]
In one of your videos where you converted an existing code-base to use StateAdapt, it was a dashboard I believe and there was a state called `UserSettings` I think. And one thing that I saw you do was manually subscribe to the store, because, you explained, if no component subscribes to the store then the state kind of disappears which means you'd lose whatever changes the user had applied to their settings.

And that made me think that we'd need something akin to RxJS's `tap()`, because then you could just persist the `UserSettings` object to `sessionStorage` and then you wouldn't need to manually forever-subscribe (which I personally don't find very elegant).

Now of course I know you can do `userSettings.state$.pipe(tap(...))`, but that's not exactly what I mean in this case. Because that puts the responsibility of adding that `tap` into the hands of the consumer of the store. While in reality, you'd want to put that responsibility in the file that created the store.

And that's currently not possible with how StateAdapt works. Well I guess one could do this of course:

```ts
// lib/stores.ts

export const userSettings = adapt(/* ... */)

userSettings.state$.pipe(
    tap(state =>
        sessionStorage.set("user-settings", JSON.stringify(state))
    )
)
.subscribe(() => {})
```

But then you're still manually creating a forever-subscription. And that doesn't yet solve the problem of also restoring the state from `sessionStorage` whenever it goes from 0 subscribers to 1 subscriber.

So, it's currently just impossible to do this with StateAdapt, but I believe that this is a very reasonable feature to want, no?

I can't explain it very well, but my intuition tells me that a middleware type of solution would be the best way to solve this cleanly. Because then you could hopefully just write a function called `persistStore` that you could call like `const userSettings = persistStore("user-settings", adapt(...))`.

What do you think?

**Comment by mfp22** [comment]
I don't think manually subscribing is that bad if you really want to keep something in state. I thought about adding an extra option to the api, but it would be annoying to maintain. I think the general way to solve what I was doing in that video is to allow passing share options, which is good for many reasons, such as keeping state for a specified amount of time before clearing it out.

But I still shy away from middleware. I generally don't like middleware patterns. Eventually maybe middleware could be implemented for StateAdapt. But I am really fighting to keep StateAdapt simple.

I had plans to add a thing called 'sinks' that would allow you to declare side effects with the store. It would be another option like 'sources', and that's how I plan on implementing localStorage and sessionStorage effects. But it would still require a manual subscription.

**Comment by Evertt** [comment]
Yeah I guess adding sinks would also solve this problem. Still not a fan of the manual forever-subscription though. I don't like the idea that I'm keeping something in memory that doesn't strictly need to be there. Feels like a memory leak waiting to happen.

### edit

Oh wait, I have an idea. What if `adapt()` doesn't just allow a simple value as it initial value, but also allows a factory function to generate the initial value? Then I could do this:

```ts
const DEFAULT_SETTINGS = {
    // ...
}

const userSettings = adapt(
    () => JSON.parse(sessionStorage.get("user-settings") ?? DEFAULT_SETTINGS),
    {
        // regular adapter stuff
        sinks: {
            persist: settings => sessionStorage.set("user-settings", JSON.stringify(settings)),
        }
    }
)
```

With that setup it wouldn't need a manual forever-subscription, because then I could trust that on every new subscription the initial state would be set from whatever is stored in `sessionStorage`.
